cpp/src/arrow/compute/row/grouper_benchmark.cc

```diff
@@ -0,0 +1,127 @@
+      });
+  // clang-format on
+
+  std::vector<Datum> values;
+  values.resize(num_types);
+  for (int i = 0; i < num_types; ++i) {
+    auto field = ::arrow::field("", types[i], metadata);
+    values[i] = rng.ArrayOf(*field, num_rows, alignment, memory_pool);
+  }
+
+  return ExecBatch(std::move(values), num_rows);
+}
+
+static void GrouperBenchmark(benchmark::State& state, const ExecSpan& span,
+                             ExecContext* ctx = nullptr) {
+  for (auto _ : state) {
+    ASSIGN_OR_ABORT(auto grouper, Grouper::Make(span.GetTypes(), ctx));
+    for (int i = 0; i < kRound; ++i) {
+      ASSIGN_OR_ABORT(auto group_ids, grouper->Consume(span));
+    }
+  }
+
+  state.SetItemsProcessed(state.iterations() * kRound * span.length);
+}
+
+static void GrouperWithMultiTypes(benchmark::State& state, const DataTypeVector& types) {
+  auto ctx = default_exec_context();
+
+  RegressionArgs args(state, false);
+  const int64_t num_rows = args.size;
+  const double null_proportion = args.null_proportion;
+
+  auto exec_batch = MakeRandomExecBatch(types, num_rows, null_proportion,
+                                        kDefaultBufferAlignment, ctx->memory_pool());
+  ExecSpan exec_span(exec_batch);
+  ASSIGN_OR_ABORT(auto grouper, Grouper::Make(exec_span.GetTypes(), ctx));
+  GrouperBenchmark(state, exec_span, ctx);
+}
+
+void SetArgs(benchmark::internal::Benchmark* bench) {
+  BenchmarkSetArgsWithSizes(bench, {1 << 10, 1 << 12});
+}
+
+// basic type column
+BENCHMARK_CAPTURE(GrouperWithMultiTypes, "{boolean}", {boolean()})->Apply(SetArgs);
+BENCHMARK_CAPTURE(GrouperWithMultiTypes, "{int32}", {int32()})->Apply(SetArgs);
+BENCHMARK_CAPTURE(GrouperWithMultiTypes, "{int64}", {int64()})->Apply(SetArgs);
+BENCHMARK_CAPTURE(GrouperWithMultiTypes, "{utf8}", {utf8()})->Apply(SetArgs);
+BENCHMARK_CAPTURE(GrouperWithMultiTypes, "{fixed_size_binary(128)}",
+                  {fixed_size_binary(128)})
+    ->Apply(SetArgs);
+
+// multi types' columns
+BENCHMARK_CAPTURE(GrouperWithMultiTypes, "{boolean, utf8}", {boolean(), utf8()})
+    ->Apply(SetArgs);
+BENCHMARK_CAPTURE(GrouperWithMultiTypes, "{int32, int32}", {int32(), int32()})
+    ->Apply(SetArgs);
+BENCHMARK_CAPTURE(GrouperWithMultiTypes, "{int32, int64}", {int32(), int64()})
+    ->Apply(SetArgs);
+BENCHMARK_CAPTURE(GrouperWithMultiTypes, "{boolean, int64, utf8}",
+                  {boolean(), int64(), utf8()})
+    ->Apply(SetArgs);
+
+// multi types' columns with column resorted
+BENCHMARK_CAPTURE(GrouperWithMultiTypes, "{int32, boolean, utf8}",
+                  {int32(), boolean(), utf8()})
+    ->Apply(SetArgs);
+BENCHMARK_CAPTURE(GrouperWithMultiTypes, "{int32, int64, boolean, utf8}",
+                  {int32(), int64(), boolean(), utf8()})
+    ->Apply(SetArgs);
+BENCHMARK_CAPTURE(GrouperWithMultiTypes,
+                  "{utf8, int32, int64, fixed_size_binary(128), boolean}",
+                  {utf8(), int32(), int64(), fixed_size_binary(128), boolean()})
+    ->Apply(SetArgs);
```

Comment:
Do we really need to benchmark so many different type combinations? Do these benchmarks exercise some particular edge cases?

Reply:
My concern is that we don't want to grow the runtime of continuous benchmarking runs gratuitously, so benchmark variations like these should be motivated by a particular concern.

Reply:
These combinations are essentially to test the impact of `RowTableEncoder`'s sorting function on input columns on the performance of `CompareColumnsToRows`. `CompareColumnsToRows` is a very low-level comparison operation, and it is necessary to ensure that the comparison performance after rearrangement will not fall back in the most extreme cases .

For comparison, the test CompareColumnsToRows in this case(https://github.com/apache/arrow/pull/40998#issuecomment-2039204161) uses unsorted columns for comparison, and the performance drops by an order of magnitude compared to rearrangement.

Reply:
Ok, I'll trust you that this can be useful. Thanks!